File: core/src/main/java/com/lomekwi/cave/ui/Root.java
package com.lomekwi.cave.ui;

import com.badlogic.gdx.ApplicationListener;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Input;
import com.badlogic.gdx.InputMultiplexer;
import com.badlogic.gdx.InputProcessor;
import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.Pixmap;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.BitmapFont;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.badlogic.gdx.graphics.g2d.freetype.FreeTypeFontGenerator;
import com.badlogic.gdx.scenes.scene2d.Actor;
import com.badlogic.gdx.scenes.scene2d.InputEvent;
import com.badlogic.gdx.scenes.scene2d.InputListener;
import com.badlogic.gdx.scenes.scene2d.Stage;
import com.badlogic.gdx.scenes.scene2d.ui.CheckBox;
import com.badlogic.gdx.scenes.scene2d.ui.Label;
import com.badlogic.gdx.scenes.scene2d.ui.Skin;
import com.badlogic.gdx.scenes.scene2d.ui.TextButton;
import com.badlogic.gdx.scenes.scene2d.ui.TextField;
import com.badlogic.gdx.scenes.scene2d.ui.Window;
import com.badlogic.gdx.scenes.scene2d.utils.DragAndDrop;
import com.badlogic.gdx.utils.ScreenUtils;
import com.badlogic.gdx.utils.viewport.ScreenViewport;
import com.kotcrab.vis.ui.VisUI;
import com.kotcrab.vis.ui.util.ToastManager;
import com.kotcrab.vis.ui.widget.LinkLabel;
import com.kotcrab.vis.ui.widget.Menu;
import com.kotcrab.vis.ui.widget.MenuItem;
import com.kotcrab.vis.ui.widget.VisTable;
import com.kotcrab.vis.ui.widget.VisTextButton;
import com.kotcrab.vis.ui.widget.VisTextField;
import com.kotcrab.vis.ui.widget.tabbedpane.TabbedPane;
import com.lomekwi.cave.Main;
import com.lomekwi.cave.project.Project;
import com.lomekwi.cave.ui.editpanel.EditPanel;
import com.lomekwi.cave.ui.editpanel.tlarea.TlGroup;
import com.lomekwi.cave.ui.tabs.ProjectTab;
import com.lomekwi.cave.ui.tabs.TopTabbedPane;
import com.lomekwi.cave.ui.topbar.TopBar;
import com.lomekwi.cave.app.App;

import space.earlygrey.shapedrawer.ShapeDrawer;

public class Root implements ApplicationListener {
    private final Main main;

    private Stage stage;

    private ToastManager toastManager;

    private VisTable mainLayout;
    private VisTable majorArea;

    private TopBar topBar;
    private TopTabbedPane tabbedPane;

    private DragAndDrop dragAndDrop;

    private ShapeDrawer shapeDrawer;

    private FreeTypeFontGenerator generator;


    public Root(Main main) {
        this.main = main;
        App.root = this;
    }


    @Override
    public void create() {
        dragAndDrop=new DragAndDrop();

        InputMultiplexer multiplexer = new InputMultiplexer();

        multiplexer.addProcessor(0, new InputProcessor() {
            @Override
            public boolean keyDown(int keycode) {
                // TopActions that don't need a project
                if (App.shortcutManager.isActive(TopBar.TopActions.NEW)) {
                    topBar.performNew();
                    return true;
                }
                if (App.shortcutManager.isActive(TopBar.TopActions.OPEN)) {
                    topBar.performOpen();
                    return true;
                }
                if (App.shortcutManager.isActive(TopBar.TopActions.CLOSE)) {
                    topBar.performClose();
                    return true;
                }

                Project project = getFrontendProject();
                if (project == null || project.undoManager == null) return false;
                EditPanel ep = getFrontendEditPanel();
                // TopActions that need a project
                if (App.shortcutManager.isActive(TopBar.TopActions.SAVE)) {
                    topBar.performSave();
                    return true;
                }
                if (App.shortcutManager.isActive(TopBar.TopActions.SAVE_AS)) {
                    topBar.performSaveAs();
                    return true;
                }

                if (isTextInputFocused()) return false;

                // Undo / Redo
                if (App.shortcutManager.isActive(TlGroup.Actions.UNDO)) {
                    project.undoManager.undo();
                    if (ep != null) ep.getTlGroup().markTimelineDirty();
                    return true;
                }
                if (App.shortcutManager.isActive(TlGroup.Actions.REDO)) {
                    project.undoManager.redo();
                    if (ep != null) ep.getTlGroup().markTimelineDirty();
                    return true;
                }

                // TlGroup actions
                TlGroup tlGroup = (ep != null) ? ep.getTlGroup() : null;
                if (tlGroup != null) {
                    if (App.shortcutManager.isActive(TlGroup.Actions.PLAY_PAUSE)) {
                        project.playhead.setPlaying(!project.playhead.isPlaying());
                        return true;
                    }
                    if (App.shortcutManager.isActive(TlGroup.Actions.SPLIT)) {
                        tlGroup.performSplit();
                        return true;
                    }
                    if (App.shortcutManager.isActive(TlGroup.Actions.DELETE)) {
                        tlGroup.performDelete();
                        return true;
                    }
                    if (App.shortcutManager.isActive(TlGroup.Actions.GROUP)) {
                        tlGroup.performGroup();
                        return true;
                    }
                }

                return false;
            }

            @Override
            public boolean keyUp(int keycode) { return false; }

            @Override
            public boolean keyTyped(char character) { return false; }

            @Override
            public boolean touchDown(int screenX, int screenY, int pointer, int button) { return false; }

            @Override
            public boolean touchUp(int screenX, int screenY, int pointer, int button) { return false; }

            @Override
            public boolean touchDragged(int screenX, int screenY, int pointer) { return false; }

            @Override
            public boolean mouseMoved(int screenX, int screenY) { return false; }

            @Override
            public boolean scrolled(float amountX, float amountY) { return false; }

            @Override
            public boolean touchCancelled(int screenX, int screenY, int pointer, int button) { return false; }
        });

        VisUI.load(injectChineseFont(VisUI.SkinScale.X2));
        stage = new Stage(new ScreenViewport());
        stage.addCaptureListener(new InputListener() {
            @Override
            public boolean touchDown(InputEvent event, float x, float y, int pointer, int button) {
                if (button != Input.Buttons.LEFT) return false;
                Actor target = event.getTarget();
                while (target != null) {
                    if (target instanceof Focusable) {
                        Actor focusTarget = target;
                        Gdx.app.postRunnable(() -> {
                            stage.setKeyboardFocus(focusTarget);
                            stage.setScrollFocus(focusTarget);
                        });
                        return false;
                    }
                    target = target.getParent();
                }
                return false;
            }
        });
        multiplexer.addProcessor(stage);
        Gdx.input.setInputProcessor(multiplexer);

        toastManager=new ToastManager(stage);

        Pixmap white = new Pixmap(1, 1, Pixmap.Format.RGBA8888);
        white.setColor(Color.WHITE);
        white.fill();
        TextureRegion region = new TextureRegion(new Texture(white));
        this.shapeDrawer = new ShapeDrawer(stage.getBatch(), region);
        white.dispose();

        // 初始化UI组件
        mainLayout = new VisTable();
        mainLayout.setBackground("window-bg");
        mainLayout.setFillParent(true);

        topBar = new TopBar();
        mainLayout.top();
        mainLayout.add(topBar.getTable()).fillX().top().row();

        tabbedPane = new TopTabbedPane();
        mainLayout.add(tabbedPane.getTable()).fillX().top().row();
        majorArea=new VisTable();
        mainLayout.add(majorArea).fill().expand().row();

        stage.addActor(mainLayout);

        registerDefaultShortcuts();

        toastManager.toFront();

        stage.setDebugAll(false);
    }

    @Override
    public void render() {
        ScreenUtils.clear(Color.BLACK);
        stage.act(Math.min(Gdx.graphics.getDeltaTime(), 1 / 30f));
        stage.draw();
    }

    @Override
    public void pause() {}

    @Override
    public void resume() {}

    @Override
    public void resize(int width, int height) {
        if (stage != null) {
            stage.getViewport().update(width, height, true);
            toastManager.resize();
        }
    }

    @Override
    public void dispose() {
        if (stage != null) stage.dispose();
        VisUI.dispose();
        generator.dispose();
    }
    public Stage getStage() {
        return stage;
    }

    public boolean isTextInputFocused() {
        if (stage == null) return false;
        var focus = stage.getKeyboardFocus();
        return focus instanceof TextField || focus instanceof VisTextField;
    }

    public VisTable getMainLayout() {
        return mainLayout;
    }
    public VisTable getMajorArea() {
        return majorArea;
    }
    private Skin injectChineseFont(VisUI.SkinScale scale) {
        Skin skin = new Skin(scale.getSkinFile());

        generator = new FreeTypeFontGenerator(Gdx.files.internal("font/noto.otf"));

        FreeTypeFontGenerator.FreeTypeFontParameter param =
            new FreeTypeFontGenerator.FreeTypeFontParameter();

        param.size = 22;
        param.incremental = true;

        BitmapFont font = generator.generateFont(param);

        skin.add("default-font", font);

        skin.get("default", Label.LabelStyle.class).font = font;
        skin.get("default", TextButton.TextButtonStyle.class).font = font;
        skin.get("default", TextField.TextFieldStyle.class).font = font;
        skin.get("default", CheckBox.CheckBoxStyle.class).font = font;
        skin.get("default", MenuItem.MenuItemStyle.class).font = font;
        skin.get("default", Menu.MenuStyle.class).openButtonStyle.font = font;
        skin.get("default", TabbedPane.TabbedPaneStyle.class).buttonStyle.font = font;
        skin.get("default", Window.WindowStyle.class).titleFont = font;
        skin.get("default", LinkLabel.LinkLabelStyle.class).font = font;
        skin.get("default", VisTextButton.VisTextButtonStyle.class).font = font;
        skin.get("default", VisTextField.VisTextFieldStyle.class).font = font;
       return skin;
    }

    public Project getFrontendProject() {
        if (tabbedPane.getActiveTab() instanceof ProjectTab) {
            return ((ProjectTab) tabbedPane.getActiveTab()).getProject();
        }else{
            return null;
        }
    }

    public EditPanel getFrontendEditPanel() {
        if (tabbedPane.getActiveTab() instanceof ProjectTab) {
            return ((ProjectTab) tabbedPane.getActiveTab()).getEditPanel();
        }else{
            return null;
        }
    }

    private void registerDefaultShortcuts() {
        for (TlGroup.Actions action : TlGroup.Actions.values()) {
            App.shortcutManager.register(action, action.defaultKeys());
        }
        for (TopBar.TopActions action : TopBar.TopActions.values()) {
            App.shortcutManager.register(action, action.defaultKeys());
        }
        App.shortcutManager.load();
        topBar.applyCustomShortcuts();
    }

    public DragAndDrop getDragAndDrop() {
        return dragAndDrop;
    }

    public ShapeDrawer getShapeDrawer() {
        return shapeDrawer;
    }

    public ToastManager getToastManager() {
        return toastManager;
    }
}

File: core/src/main/scala/com/lomekwi/cave/ui/Root.scala
package com.lomekwi.cave.ui

import com.badlogic.gdx.ApplicationListener
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.Input
import com.badlogic.gdx.InputMultiplexer
import com.badlogic.gdx.InputProcessor
import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.Pixmap
import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.graphics.g2d.BitmapFont
import com.badlogic.gdx.graphics.g2d.TextureRegion
import com.badlogic.gdx.graphics.g2d.freetype.FreeTypeFontGenerator
import com.badlogic.gdx.scenes.scene2d.Actor
import com.badlogic.gdx.scenes.scene2d.InputEvent
import com.badlogic.gdx.scenes.scene2d.InputListener
import com.badlogic.gdx.scenes.scene2d.Stage
import com.badlogic.gdx.scenes.scene2d.ui.CheckBox
import com.badlogic.gdx.scenes.scene2d.ui.Container
import com.badlogic.gdx.scenes.scene2d.ui.Label
import com.badlogic.gdx.scenes.scene2d.ui.List
import com.badlogic.gdx.scenes.scene2d.ui.SelectBox
import com.badlogic.gdx.scenes.scene2d.ui.Skin
import com.badlogic.gdx.scenes.scene2d.ui.Table
import com.badlogic.gdx.scenes.scene2d.ui.TextButton
import com.badlogic.gdx.scenes.scene2d.ui.TextField
import com.badlogic.gdx.scenes.scene2d.ui.Window
import com.badlogic.gdx.scenes.scene2d.utils.DragAndDrop
import com.badlogic.gdx.utils.ScreenUtils
import com.badlogic.gdx.utils.viewport.ScreenViewport
import com.kotcrab.vis.ui.VisUI
import com.kotcrab.vis.ui.util.ToastManager
import com.kotcrab.vis.ui.widget.LinkLabel
import com.kotcrab.vis.ui.widget.Menu
import com.kotcrab.vis.ui.widget.MenuItem
import com.kotcrab.vis.ui.widget.VisTable
import com.kotcrab.vis.ui.widget.VisTextButton
import com.kotcrab.vis.ui.widget.VisTextField
import com.kotcrab.vis.ui.widget.tabbedpane.TabbedPane
import com.lomekwi.cave.project.Project
import com.lomekwi.cave.ui.editpanel.EditPanel
import com.lomekwi.cave.ui.editpanel.tlarea.TimelineView
import com.lomekwi.cave.ui.tabs.app.ProjectTab
import com.lomekwi.cave.ui.tabs.app.TopTabbedPane
import com.lomekwi.cave.ui.topbar.TopBar
import com.lomekwi.cave.app.App

import space.earlygrey.shapedrawer.ShapeDrawer
import scala.compiletime.uninitialized

class Root extends ApplicationListener {
  private var stage: Stage = uninitialized

  private var toastManager: ToastManager = uninitialized

  private var mainLayout: VisTable = uninitialized
  private var majorArea: Container[Table] = uninitialized

  private var topBar: TopBar = uninitialized
  private var tabbedPane: TopTabbedPane = uninitialized

  private var dragAndDrop: DragAndDrop = uninitialized

  private var shapeDrawer: ShapeDrawer = uninitialized

  private var generator: FreeTypeFontGenerator = uninitialized

  App.root = this

  override def create(): Unit = {
    dragAndDrop = new DragAndDrop()

    val multiplexer: InputMultiplexer = new InputMultiplexer()

    multiplexer.addProcessor(0, new InputProcessor {
      override def keyDown(keycode: Int): Boolean = {
        val project: Project = getFrontendProject
        val ep: EditPanel = getFrontendEditPanel
        // 无需项目的 TopActions
        if (App.shortcutManager.isActive(TopBar.TopActions.NEW)) {
          topBar.performNew()
          true
        } else if (App.shortcutManager.isActive(TopBar.TopActions.OPEN)) {
          topBar.performOpen()
          true
        } else if (App.shortcutManager.isActive(TopBar.TopActions.CLOSE)) {
          topBar.performClose()
          true
        } else if (project == null || project.undoManager == null) {
          false
        } else if (App.shortcutManager.isActive(TopBar.TopActions.SAVE)) {
          // 需要项目的 TopActions
          topBar.performSave()
          true
        } else if (App.shortcutManager.isActive(TopBar.TopActions.SAVE_AS)) {
          topBar.performSaveAs()
          true
        } else if (isTextInputFocused) {
          false
        } else if (App.shortcutManager.isActive(TimelineView.Actions.COPY)) {
          // 复制（全局，无需项目）
          App.copyManager.copy()
          true
        } else if (App.shortcutManager.isActive(TimelineView.Actions.UNDO)) {
          // 撤销 / 重做
          project.undoManager.undo()
          if (ep != null) ep.getTimelineView.markTimelineDirty()
          true
        } else if (App.shortcutManager.isActive(TimelineView.Actions.REDO)) {
          project.undoManager.redo()
          if (ep != null) ep.getTimelineView.markTimelineDirty()
          true
        } else {
          false
        }
      }

      override def keyUp(keycode: Int): Boolean = false

      override def keyTyped(character: Char): Boolean = false

      override def touchDown(screenX: Int, screenY: Int, pointer: Int, button: Int): Boolean = false

      override def touchUp(screenX: Int, screenY: Int, pointer: Int, button: Int): Boolean = false

      override def touchDragged(screenX: Int, screenY: Int, pointer: Int): Boolean = false

      override def mouseMoved(screenX: Int, screenY: Int): Boolean = false

      override def scrolled(amountX: Float, amountY: Float): Boolean = false

      override def touchCancelled(screenX: Int, screenY: Int, pointer: Int, button: Int): Boolean = false
    })

    VisUI.load(injectChineseFont(VisUI.SkinScale.X2))
    stage = new Stage(new ScreenViewport())
    stage.addCaptureListener(new InputListener {
      override def touchDown(event: InputEvent, x: Float, y: Float, pointer: Int, button: Int): Boolean = {
        // 文本输入控件自己会接管键盘焦点，这里再抢一次会让它立刻失焦
        if (button == Input.Buttons.LEFT && !TextInputs.contains(event.getTarget)) {
          var target: Actor = event.getTarget
          while (target != null && !target.isInstanceOf[Focusable]) {
            target = target.getParent
          }
          if (target != null) {
            val focusTarget: Actor = target
            Gdx.app.postRunnable(() => {
              stage.setKeyboardFocus(focusTarget)
              stage.setScrollFocus(focusTarget)
            })
          }
        }
        false
      }
    })
    multiplexer.addProcessor(stage)
    Gdx.input.setInputProcessor(multiplexer)

    toastManager = new ToastManager(stage)

    val white: Pixmap = new Pixmap(1, 1, Pixmap.Format.RGBA8888)
    white.setColor(Color.WHITE)
    white.fill()
    val region: TextureRegion = new TextureRegion(new Texture(white))
    this.shapeDrawer = new ShapeDrawer(stage.getBatch, region)
    white.dispose()

    mainLayout = new VisTable()
    mainLayout.setBackground("window-bg")
    mainLayout.setFillParent(true)

    topBar = new TopBar()
    mainLayout.top()
    mainLayout.add(topBar.getTable).fillX().top().row()

    tabbedPane = new TopTabbedPane()
    mainLayout.add(tabbedPane.getTable).fillX().top().row()
    tabbedPane.refreshVisibility()
    majorArea = new Container[Table] {
      override def getMinHeight: Float = {
        0
      }
      override def getMinWidth: Float = {
        0
      }
      override def getPrefHeight: Float = {
        0
      }
      override def getPrefWidth: Float = {
        0
      }
    }
    majorArea.fill()
    mainLayout.add(majorArea).fill().expand().row()

    stage.addActor(mainLayout)

    registerDefaultShortcuts()

    App.appEventBus.register(App.copyManager)

    toastManager.toFront()

    stage.setDebugAll(false)
  }

  override def render(): Unit = {
    ScreenUtils.clear(Color.BLACK)
    stage.act(Math.min(Gdx.graphics.getDeltaTime, 1 / 30f))
    stage.draw()
  }

  override def pause(): Unit = {}

  override def resume(): Unit = {}

  override def resize(width: Int, height: Int): Unit = {
    if (stage != null) {
      stage.getViewport.update(width, height, true)
      toastManager.resize()
    }
  }

  override def dispose(): Unit = {
    if (stage != null) stage.dispose()
    VisUI.dispose()
    generator.dispose()
  }
  def getStage: Stage = {
    stage
  }

  def isTextInputFocused: Boolean = {
    if (stage == null) {
      false
    } else {
      val focus = stage.getKeyboardFocus
      focus.isInstanceOf[TextField] || focus.isInstanceOf[VisTextField]
    }
  }

  def getMainLayout: VisTable = {
    mainLayout
  }
  def getMajorArea: Container[Table] = {
    majorArea
  }
  def getTabbedPane: TopTabbedPane = {
    tabbedPane
  }
  private def injectChineseFont(scale: VisUI.SkinScale): Skin = {
    val skin: Skin = new Skin(scale.getSkinFile)

    generator = new FreeTypeFontGenerator(Gdx.files.internal("font/noto.otf"))

    val param: FreeTypeFontGenerator.FreeTypeFontParameter =
      new FreeTypeFontGenerator.FreeTypeFontParameter()

    param.size = 22
    param.incremental = true

    val font: BitmapFont = generator.generateFont(param)

    skin.add("default-font", font)

    skin.get("default", classOf[Label.LabelStyle]).font = font
    skin.get("default", classOf[TextButton.TextButtonStyle]).font = font
    skin.get("default", classOf[TextField.TextFieldStyle]).font = font
    skin.get("default", classOf[CheckBox.CheckBoxStyle]).font = font
    skin.get("default", classOf[MenuItem.MenuItemStyle]).font = font
    skin.get("default", classOf[Menu.MenuStyle]).openButtonStyle.font = font
    skin.get("default", classOf[TabbedPane.TabbedPaneStyle]).buttonStyle.font = font
    skin.get("default", classOf[Window.WindowStyle]).titleFont = font
    skin.get("default", classOf[LinkLabel.LinkLabelStyle]).font = font
    skin.get("default", classOf[VisTextButton.VisTextButtonStyle]).font = font
    skin.get("default", classOf[VisTextField.VisTextFieldStyle]).font = font
    val selectStyle: SelectBox.SelectBoxStyle = skin.get("default", classOf[SelectBox.SelectBoxStyle])
    selectStyle.font = font
    selectStyle.listStyle.font = font
    skin.get("default", classOf[List.ListStyle]).font = font
    skin
  }

  def getFrontendProject: Project = tabbedPane.getActiveTab match {
    case tab: ProjectTab => tab.getProject
    case _ => null
  }

  def getFrontendEditPanel: EditPanel = tabbedPane.getActiveTab match {
    case tab: ProjectTab => tab.getEditPanel
    case _ => null
  }

  private def registerDefaultShortcuts(): Unit = {
    for (action <- TimelineView.Actions.values) {
      App.shortcutManager.register(action, action.defaultKeys()*)
    }
    for (action <- TopBar.TopActions.values) {
      App.shortcutManager.register(action, action.defaultKeys()*)
    }
    App.shortcutManager.load()
    topBar.applyCustomShortcuts()
  }

  def getDragAndDrop: DragAndDrop = {
    dragAndDrop
  }

  def getShapeDrawer: ShapeDrawer = {
    shapeDrawer
  }

  def getToastManager: ToastManager = {
    toastManager
  }
}
